proposals/p5434.md

```diff
@@ -0,0 +1,1331 @@
+fn Invalid1(...) -> (bool, ->let f32);
+
+// ❌ Invalid, can only specify return category
+// on innermost `->`.
+fn Invalid2(...) -> var (-> bool, -> f32);
+```
+
+To nest, use the default `->` in the outer form, as in:
+
+```carbon
+fn BracesInParens(...)
+    -> (->{-> .a: bool, ->let .b: f32}, ->var C, -ref i32);
+
+let ({.a = var x: bool, .b = let y: f32},
+     var c: C, ref d: i32) = BracesInParens();
+// or without renaming:
+let ({var a: bool, let b: f32},
+     var c: C, ref d: i32) = BracesInParens();
+
+fn ParensInBraces(...)
+    -> {-> .a: bool,
+        -> .b: (->let f32, ->var C),
+        ->ref .c: i32};
+
+let {a: bool,
+     .b = (x: f32, var y: C),
+     ref c: i32} = ParensInBraces(...);
+```
+
+This feature is intended to support cases like `enumerate` that will want to
+return a value for the index but a reference to the element of the sequence
+being enumerated.
+
+### Nested binding patterns
+
+Since a `ref` binding may only bind to a durable reference expression, it can't
+be used to bind the result of a function returning an initializing expression.
+However, if the initializing expression is bound to a `var`, any nested patterns
+are reference binding patterns bound to the subobject, following
+[proposal #5164: "Updates to pattern matching for objects"](https://github.com/carbon-language/carbon-lang/pull/5164).
+
+For example:
+
+```carbon
+fn F() -> (bool, (C, i32));
+let (b: bool, var (c: C, i: i32)) = F();
+```
+
+is equivalent to:
+
+```carbon
+fn F() -> (bool, (C, i32));
+let (b: bool, var v: (C, i32)) = F();
+let ref c: C = v.0;
+let ref i: i32 = v.1;
+```
+
+### `bound` parameters
+
+It is an error to return a reference to a temporary object that won't live once
+the function returns, even if it is a parameter marked `bound`.
+
+```carbon
+fn Invalid1() -> i32* {
+  var x: i32 = 4;
+  // ❌ Invalid.
+  return &x;
+}
+
+fn Invalid2(bound x: i32) -> ref i32 {
+  var y: i32 = x;
+  // ❌ Invalid.
+  return y;
+}
+
+// ✅ Valid
+fn Valid1(bound p: i32*) -> ref i32 {
+  return *p;
+}
+
+fn Invalid3(bound var x: i32) -> ref i32 {
+  // ❌ Invalid: lifetime of `var` parameter
+  // ends when function returns.
+  return x;
+}
+
+class ReturnMember {
+  // ✅ Valid
+  fn ValidRef[bound ref self: Self]() -> ref i32 {
+    return self.m;
+  }
+
+  // ❌ Invalid: can't return reference to value.
+  fn InvalidLet[bound self: Self]() -> ref i32 {
+    return self.m;
+  }
+
+  // ❌ Invalid: `var self` lifetime ends.
+  fn InvalidVar[bound var self: Self]()
+      -> ref i32 { return self.m; }
+
+  var m: i32;
+}
+
+class DerefPointerMember {
+  // ✅ Valid
+  fn ValidRef[bound ref self: Self]() -> ref i32 {
+    return *self.pm;
+  }
+
+  // ✅ Valid
+  fn ValidLet[bound self: Self]() -> ref i32 {
+    return *self.pm;
+  }
+
+  // ✅ Valid
+  fn ValidVar[bound var self: Self]()
+      -> ref i32 { return *self.pm; }
+
+  var pm: i32*;
+}
+```
+
+The `bound` parameters must include everything referenced by the return, but
```

Comment:
That doesn't seem quite right, since the return can reference global variables. The phrasing is also a little awkward. How about something like "The returned expression for a `-> ref` cannot reference any local names not marked with `bound`"?